File: packages/coast_audio_miniaudio/mabridge/src/mab_peaking_eq_filter.c
#define MINIAUDIO_IMPLEMENTATION
#include "miniaudio.h"

#include "mab_types.h"
#include "mab_peaking_eq_filter.h"

mab_peaking_eq_filter_config mab_peaking_eq_filter_config_init(mab_format format, u_int32_t sampleRate, u_int32_t channels, double gainDb, double q, double frequency)
{
  mab_peaking_eq_filter_config config = {
      .format = format,
      .sampleRate = sampleRate,
      .channels = channels,
      .gainDb = gainDb,
      .q = q,
      .frequency = frequency,
  };
  return config;
}

mab_result mab_peaking_eq_filter_init(mab_peaking_eq_filter* pEQ, mab_peaking_eq_filter_config config)
{
  ma_peak2_config maConfig = ma_peak2_config_init(mab_cast(ma_format, config.format), config.channels, config.sampleRate, config.gainDb, config.q, config.frequency);
  ma_peak2* pData = MAB_MALLOC(sizeof(ma_peak2));
  ma_result result = ma_peak2_init(&maConfig, NULL, pData);
  if (result != MA_SUCCESS) {
    MAB_FREE(pData);
    return result;
  }

  pEQ->pData = pData;
  return MA_SUCCESS;
}

mab_result mab_peaking_eq_filter_process(mab_peaking_eq_filter* pEQ, void* pFramesOut, const void* pFramesIn, u_int64_t frameCount)
{
  ma_peak2 *pData = (ma_peak2*)pEQ->pData;
  return ma_peak2_process_pcm_frames(pData, pFramesOut, pFramesIn, frameCount);
}

mab_result mab_peaking_eq_filter_reinit(mab_peaking_eq_filter* pEQ, mab_peaking_eq_filter_config config)
{
  ma_peak2 *pData = (ma_peak2*)pEQ->pData;
  ma_peak2_config maConfig = ma_peak2_config_init(mab_cast(ma_format, config.format), config.channels, config.sampleRate, config.gainDb, config.q, config.frequency);
  return ma_peak2_reinit(&maConfig, pData);
}

u_int32_t mab_peaking_eq_filter_get_latency(mab_peaking_eq_filter* pEQ)
{
  ma_peak2 *pData = (ma_peak2*)pEQ->pData;
  return ma_peak2_get_latency(pData);
}

void mab_peaking_eq_filter_uninit(mab_peaking_eq_filter* pEQ)
{
  ma_peak2 *pData = (ma_peak2*)pEQ->pData;
  ma_peak2_uninit(pData, NULL);
  MAB_FREE(pData);
  pEQ->pData = NULL;
}

File: packages/coast_audio_miniaudio/lib/src/ma_bridge/filters/mab_peaking_eq_filter.dart
import 'dart:ffi';

import 'package:coast_audio/coast_audio.dart';
import 'package:coast_audio_miniaudio/generated/ma_bridge_bindings.dart';
import 'package:coast_audio_miniaudio/src/ma_bridge/filters/mab_filter_base.dart';
import 'package:coast_audio_miniaudio/src/ma_extension.dart';

/// High shelf filter. It uses ma_peak2 implementations.
class MabPeakingEqFilter extends MabFilterBase {
  MabPeakingEqFilter({
    required this.format,
    required double gainDb,
    required double q,
    required double frequency,
    super.memory,
  })  : _gainDb = gainDb,
        _q = q,
        _frequency = frequency {
    final config = library.mab_peaking_eq_filter_config_init(format.sampleFormat.mabFormat.value, format.sampleRate, format.channels, gainDb, q, frequency);
    library.mab_peaking_eq_filter_init(_pEQ, config).throwMaResultIfNeeded();
    _updateLatency();
  }

  final AudioFormat format;

  double _gainDb;
  double get gainDb => _gainDb;

  double _q;
  double get q => _q;

  double _frequency;
  double get frequency => _frequency;

  late AudioTime _latency;
  AudioTime get latency => _latency;

  late final _pEQ = allocate<mab_peaking_eq_filter>(sizeOf<mab_peaking_eq_filter>());

  @override
  void process(AudioBuffer bufferOut, AudioBuffer bufferIn) {
    assert(bufferOut.sizeInFrames >= bufferIn.sizeInFrames);
    library.mab_peaking_eq_filter_process(_pEQ, bufferOut.pBuffer.cast(), bufferIn.pBuffer.cast(), bufferIn.sizeInFrames).throwMaResultIfNeeded();
  }

  /// Reinit the filter parameters while keeping internal state.
  void update({double? gainDb, double? q, double? frequency}) {
    final config = library.mab_peaking_eq_filter_config_init(
      format.sampleFormat.mabFormat.value,
      format.sampleRate,
      format.channels,
      gainDb ?? this.gainDb,
      q ?? this.q,
      frequency ?? this.frequency,
    );
    library.mab_peaking_eq_filter_reinit(_pEQ, config).throwMaResultIfNeeded();
    _gainDb = gainDb ?? this.gainDb;
    _q = q ?? this.q;
    _frequency = frequency ?? this.frequency;
    _updateLatency();
  }

  void _updateLatency() {
    final frameCount = library.mab_peaking_eq_filter_get_latency(_pEQ);
    _latency = AudioTime.fromFrames(frames: frameCount, format: format);
  }

  @override
  void uninit() {
    library.mab_peaking_eq_filter_uninit(_pEQ);
  }
}
